sky/execution.py

```diff
@@ -161,20 +162,26 @@ def _execute(
         if dryrun.
     """
     cluster_exists = False
+    cluster_running = False
     if cluster_name is not None:
-        existing_handle = global_user_state.get_handle_from_cluster_name(
-            cluster_name)
-        cluster_exists = existing_handle is not None
+        cluster_record = global_user_state.get_cluster_from_name(cluster_name)
+        cluster_exists = cluster_record is not None
+        cluster_running = (cluster_record is not None and
```

Comment:
Hmm, this doesn't really refresh the status so the name is not 100% accurate? Do we need it, or is cluster_exists enough?

Reply:
`cluster_exists` is not enough, because to enforce autostop, we have to enforce autostop is passed for a `launch` request on a stopped cluster, while `cluster_exists` only reflects the cluster already exists not reflecting the status.

This is a compromise for the performance vs accuracy. An alternative would be asking the admin to remove this field from the argument and let them to call `sky.status(refresh=True)` for the cluster themselves.

Reply:
An admin will specify the following instead:
```python
class EnforceAutostopPolicy(sky.AdminPolicy):
    """Example policy: enforce autostop for all tasks."""

    @classmethod
    def validate_and_mutate(
            cls, user_request: sky.UserRequest) -> sky.MutatedUserRequest:
        """Enforces autostop for all tasks."""
        request_options = user_request.request_options
        # Request options is None when a task is executed with `jobs launch` or
        # `sky serve up`.
        if request_options is None:
            return sky.MutatedUserRequest(
                task=user_request.task,
                skypilot_config=user_request.skypilot_config)
        cluster_record = sky.status(request_options.cluster_name, refresh=True)
        need_autostop = False
        if not cluster_record:
            # Cluster does not exist
            need_autostop = True
        elif cluster_record[0]['status'] == sky.ClusterStatus.STOPPED:
            # Cluster is stopped
            need_autostop = True
        elif cluster_record[0]['autostop'] < 0:
            # Cluster is running but autostop is not set
            need_autostop = True

        is_setting_autostop = False
        idle_minutes_to_autostop = request_options.idle_minutes_to_autostop
        is_setting_autostop = (idle_minutes_to_autostop is not None and
                               idle_minutes_to_autostop >= 0)
        if need_autostop and not is_setting_autostop:
            raise RuntimeError('Autostop/down must be set for all clusters.')

        return sky.MutatedUserRequest(
            task=user_request.task,
            skypilot_config=user_request.skypilot_config)
```

Reply:
I updated the enforce autostop's policy to use `sky.status` instead and removed the `cluster_running` field.

Reply:
Q: what's the latency impact of this policy on `sky launch/exec`? Is it small? Do we expect users to use this example policy and be ok with the latency?

Reply:
Based on my testing, the latency is acceptable. I am adding a note box in the doc about this implication.